### ISSUE TITLE
using SVM on a dataset

### ISSUE THREAD

**Issue by parsaanooshfar-lab** [parsaanooshfar-lab/homework-machine-learning#12]
<img width="1380" height="1180" alt="Image" src="https://github.com/user-attachments/assets/26757442-7ffb-4c6e-b849-71a584efc3c8" />

# I'll train a linear SVM on a sample 2D dataset, plot the points and the decision boundary (the "SVM line"),
# and print the line equation in the form w0*x + w1*y + b = 0  and also y = mx + c
import numpy as np
import matplotlib.pyplot as plt
from sklearn import svm
from sklearn.datasets import make_blobs
from sklearn.model_selection import train_test_split

# Create a separable 2D dataset (replace this with your data if you provide it)
X, y = make_blobs(n_samples=60, centers=2, n_features=2, random_state=42, cluster_std=1.0)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

# Train a linear SVM
clf = svm.SVC(kernel='linear', C=1.0)
clf.fit(X_train, y_train)

# Get weights and intercept (for linear kernel: coef_ and intercept_ available)
w = clf.coef_[0]
b = clf.intercept_[0]

# Line equation: w0*x + w1*y + b = 0  -> y = -(w0/w1)x - b/w1 (if w1 != 0)
if abs(w[1]) < 1e-8:
    line_type = "vertical"
else:
    m = -w[0] / w[1]
    c = -b / w[1]
    line_type = f"y = {m:.4f}x + {c:.4f}"

# Print model info and accuracy
acc_train = clf.score(X_train, y_train)
acc_test = clf.score(X_test, y_test)

print("SVM linear model coefficients (w):", w)
print("Intercept (b):", b)
print("Line type / equation:", line_type)
print(f"Training accuracy: {acc_train:.3f}, Test accuracy: {acc_test:.3f}")

# Plot data and decision boundary + margins and support vectors
plt.figure(figsize=(7,6))
# scatter
plt.scatter(X_train[:,0], X_train[:,1], c=y_train)
plt.scatter(X_test[:,0], X_test[:,1], marker='x', s=80, label='test')

# create grid to evaluate model
ax = plt.gca()
xlim = ax.get_xlim()
ylim = ax.get_ylim()
xx = np.linspace(xlim[0]-1, xlim[1]+1, 200)
yy = np.linspace(ylim[0]-1, ylim[1]+1, 200)
YY, XX = np.meshgrid(yy, xx)
xy = np.vstack([XX.ravel(), YY.ravel()]).T
Z = clf.decision_function(xy).reshape(XX.shape)

# plot decision boundary and margins
contour = plt.contour(XX, YY, Z, levels=[-1.0, 0.0, 1.0], linestyles=['--','-','--'])
plt.scatter(clf.support_vectors_[:,0], clf.support_vectors_[:,1], facecolors='none', edgecolors='k', s=100, label='support vectors')

plt.title('Linear SVM decision boundary (example dataset)')
plt.xlabel('x1')
plt.ylabel('x2')
plt.legend()
plt.show()
